### PR TITLE
[Testcase Refactoring] Split test_c10d_gloo by hardware classification

### PR DESCRIPTION
**Pull request by songzibo**
## Summary

`test/distributed/test_c10d_gloo.py` contains 5 mixed CPU+CUDA test classes that were previously unclassified. This PR splits them into shared `*Base` + CPU (`GENERIC`) + `CUDA` subclasses, following the pattern in #190761 (`test_c10d_spawn_gloo`). All 16 test classes are labeled with `hw_classification`.

## Changes

- **`test/distributed/test_c10d_gloo.py`**:
  - Imported `HardwareClassification` from `torch.testing._internal.common_utils`.
  - Split 5 mixed classes into `*Base` (helpers only, not collected) + CPU (`GENERIC`) + `*CUDA` (`CUDA`):

| original class | Base (helpers) | CPU (GENERIC) | CUDA (new) |
|---|---|---|---|
| `ProcessGroupGlooTest` | `ProcessGroupGlooTestBase` | `ProcessGroupGlooTest` (48 tests) | `ProcessGroupGlooTestCUDA` (20 tests) |
| `DistributedDataParallelTest` | `DistributedDataParallelTestBase` | `DistributedDataParallelTest` (15 tests) | `DistributedDataParallelTestCUDA` (11 tests) |
| `ProcessGroupGlooLazyInitTest` | — | kept (GENERIC, inherits 48 CPU) | `ProcessGroupGlooLazyInitTestCUDA` (CUDA, inherits 20) |
| `ProcessGroupGlooFRTest` | — | kept (GENERIC, 3 own + inherits 48 CPU) | `ProcessGroupGlooFRTestCUDA` (CUDA, inherits 20) |
| `CommTest` | `CommTestBase` | `CommTest` (3 tests) | `CommTestCUDA` (6 tests) |

  - Labeled 6 pure-CPU classes `GENERIC` (no split): `RendezvousTCPTest`, `RendezvousEnvTest`, `TimeoutTest`, `ReducerTest`, `GlooProcessGroupWithDispatchedCollectivesTests`, `LargeCommTest`.

  - 37 CUDA test methods moved to the 5 new `*CUDA` classes. The `*Base` classes contain only shared fixtures and `_test_*` helpers (zero `test_*` methods, so unittest does not collect them).

## Motivation

PR #190991 established the convention that every `TestCase` subclass must carry a `hw_classification` attribute; an unclassified class is silently dropped when `--hw-classification` is passed to the test loader. The 11 classes in `test_c10d_gloo.py` had no label, so they were invisible under hardware-classification filtering.

The 5 mixed classes contained both CPU gloo tests and CUDA tests (`.cuda()`, `@skip_if_lt_x_gpu`). A single `hw_classification` label cannot represent a mixed class. Following #190761, they are split so the CPU tests are `GENERIC` (collectible under `--hw-classification GENERIC`) and the CUDA tests are `CUDA` (collectible under `--hw-classification CUDA`).

## Dependencies

None. This is a standalone label + split change with no infrastructure dependencies.

## Test Plan

- `python -m py_compile test/distributed/test_c10d_gloo.py` passes.
- `lintrunner test/distributed/test_c10d_gloo.py` passes (exit 0).
- Test count unchanged: 131 `def test_*` methods (`grep -c "def test_"`).
- 16 `hw_classification` labels (6 pure-CPU + 5 CPU split + 5 CUDA split).
- 3 `*Base` classes verified to have zero `test_*` methods via AST check.
- All `@skip_if_lt_x_gpu`, `@requires_gloo`, `@skipIfRocm` decorators preserved.
- No method bodies changed; only class location changes.

cc @fffrog @wjlFlyer @pjyFight @FuDdd @lyriexs666 @jishuangfeng @JamesD18 @JiasenTian

### CODE REVIEW

File: test/distributed/test_c10d_gloo.py [hunk 1]
```diff
@@ -1901,179 +1910,185 @@ def test_alltoall_checks(self):
...
+    def test_alltoall_multidim(self):
+        self._test_alltoall_multidim(lambda t: t.clone())
 
-        # After alltoall, rank i should have received tensor with value j from rank j
-        # in output_tensors[j]
-        for i in range(self.world_size):
-            self.assertEqual(torch.tensor([i]), output_tensors[i])
 
+class ProcessGroupGlooTestCUDA(ProcessGroupGlooTestBase):
+    hw_classification = HardwareClassification.CUDA
```
Comment (wjlFlyer) [inline test/distributed/test_c10d_gloo.py:1936]: The class tagged by `cuda` hasn't been instantiated by `instantiate_device_type_tests`.

**Comment by songzibo** [comment]
Thanks @wjlFlyer! All 5 CUDA classes are now instantiated with `instantiate_device_type_tests(only_for="cuda")` in the latest push (3825d62):

1. `ProcessGroupGlooTestCUDA` (20 test methods)
2. `DistributedDataParallelTestCUDA` (11 test methods)
3. `ProcessGroupGlooLazyInitTestCUDA` (inherits from ProcessGroupGlooTestCUDA)
4. `ProcessGroupGlooFRTestCUDA` (inherits from ProcessGroupGlooTestCUDA)
5. `CommTestCUDA` (6 test methods)

All `test_*` methods in these CUDA classes now accept a `device` parameter. All `.cuda()` calls replaced with `.to(device)` or `device=device`. Per the TEST_LINTER rule (#190173), `only_for="cuda"` is used (not `except_for`).

Verified: py_compile PASS, lintrunner PASS, 131 test methods unchanged, 0 `.cuda()` calls remaining in CUDA classes.

**Review by wjlFlyer — APPROVED** [review]
